Flag-level Action

xwjdsh opened the pull request:
<!--
  This template provides some ideas of things to include in your PR description.
  To start, try providing a short summary of your changes in the Title above.
  If a section of the PR template does not apply to this PR, then delete that section.
 -->

## What type of PR is this?

- feature

## What this PR does / why we need it:
https://github.com/urfave/cli/issues/1219
I like the idea, but there is no comment from the maintainers, I'd like to have a simple implementation first and then hear your opinions.

### Example
```go
package main

import (
	"fmt"
	"os"

	"github.com/urfave/cli/v2"
)

func main() {
	app := &cli.App{
		Action: func(c *cli.Context) error { return nil },
		Flags: []cli.Flag{
			&cli.StringFlag{
				Name: "f",
				Action: func(c *cli.Context, s string) error {
					fmt.Println(s)
					return nil
				},
			},
		},
	}

	err := app.Run(os.Args)
	if err != nil {
		panic(err)
	}
}
```

```bash
## nothing if no flag
go run main.go
## print flag value
go run main.go -f test
#> test
```

<!--
  What goal is this change working towards?
  Provide a bullet pointed summary of how each file was changed.
  Briefly explain any decisions you made with respect to the changes.
  Include anything here that you didn't include in *Release Notes*
  above, such as changes to CI or changes to internal methods.
-->

## Which issue(s) this PR fixes:

close https://github.com/urfave/cli/issues/1219
<!--
If this PR fixes one of more issues, list them here.
One line each, like so:

Fixes #123
Fixes #39
-->

dearchap commented on flag.go:
```diff
@@ -94,6 +94,7 @@ type Flag interface {
 	Apply(*flag.FlagSet) error
 	Names() []string
 	IsSet() bool
+	RunAction(*Context) error
```
This would break all existing flag implementations. So usually we wrap this into a separate interface say 

type ActionableFlag interface {
   Flag
   RunAction(*Context) error
}

dearchap commented on app.go:
```diff
@@ -506,6 +514,24 @@ func (a *App) handleExitCoder(context *Context, err error) {
 
+func runFlagActions(c *Context, fs []Flag) error {
+	for _, f := range fs {
+		isSet := false
```
check if f.(ActionableFlag) is ok and then execute. See comment below

dearchap reviewed (changes requested):
Excellent work. Thanks for completeness. Just add the new interface and then you should be all set

xwjdsh commented:
@dearchap Thank you for your review, I updated and rebased main branch.

dearchap commented on app_test.go:
```diff
@@ -2287,3 +2291,66 @@ func TestSetupInitializesOnlyNilWriters(t *testing.T) {
+	app := &App{
+		Name:   "command",
+		Writer: io.Discard,
+		Flags:  []Flag{&StringFlag{Name: "flag", Action: actionFunc}},
```
Can you add some more tests which use other types of flags ? codecov is complaining that runAction on other flags dont have any test coverage

xwjdsh replied:
No problem,  I'll fill it in.

xwjdsh replied:
Updated.

meatballhat commented:
Ack, sorry for the conflicts! I'm happy to attempt to resolve them.

meatballhat commented:
These conflicts are going to be more difficult given the flag generation code introduced in the meantime. Giving it a shot 🤞🏼

xwjdsh commented:
@meatballhat I'll fix it later, that's OK.

meatballhat commented:
@xwjdsh I'm happy to carry this work forward if you're not able ❤️

xwjdsh commented:
Sorry for being late, I updated.

Update `gopkg.in/yaml.v2` to `gopkg.in/yaml.v3` cause v2 seems have an issue that commas cannot be unmarshal correctly, I switched to v3 and the issue resolved.

meatballhat commented:
@xwjdsh I'm sorry about this frustrating feedback loop with having to approve GitHub Actions 😩

xwjdsh commented:
@meatballhat I'm a little bit confused about the CI error, my local environment is normal...

meatballhat commented:
@xwjdsh I'm sorry about this. There's something up with the way that the `go.mod` is written that seems to be upsetting for go 1.16.x. This isn't the first time this problem has come up in CI. Here's my attempted fix: https://github.com/urfave/cli/pull/1407. Once that's merged, can you incorporate latest `main`?

dearchap commented:
@xwjdsh @meatballhat Is one of you looking at the test failures and conflicts ?

meatballhat commented:
> @xwjdsh @meatballhat Is one of you looking at the test failures and conflicts ?

Not me!

xwjdsh commented:
@dearchap It's OK now.

dearchap commented on flag.go:
```diff
@@ -83,6 +83,12 @@ func (f FlagsByName) Swap(i, j int) {
+// ActionableFlag is an interface that wraps Flag interface and RunAction operation.
+type ActionableFlag interface {
+	Flag
+	RunAction(*Context) error
```
Do you think RunAction might need Flag itself as an arg ?

dearchap commented on app_test.go:
```diff
@@ -2576,3 +2582,259 @@ func TestSetupInitializesOnlyNilWriters(t *testing.T) {
+		Action: func(ctx *Context) error { return nil },
+	}
+
+	tests := []struct {
```
can you add some tests which return some error a well just for completeness ?

dearchap commented on flag_timestamp.go:
```diff
@@ -148,6 +148,15 @@ func (f *TimestampFlag) Get(ctx *Context) *time.Time {
+		return f.Action(c, c.Timestamp(f.Name))
+	}
+
+	return nil
```
Can you add a testcase for a flag which doesnt have an Action defined but is set on cmdline ?
